### ISSUE TITLE
Firefox refresh loop causing thousands of "max pending messages" errors

### ISSUE THREAD

**Issue by tomlut** [home-assistant/frontend#4404]
**Checklist:**

- [x] I updated to the latest version available
- [x] I cleared the cache of my browser

**Home Assistant release with the issue:**
0.103.5

**Last working Home Assistant release (if known):**
Not known
**UI (States or Lovelace UI?):**
Lovelace

**Browser and Operating System:**
Firefox 71.0 (64-bit)
Windows 10 Pro 64 bit 1903 build 18362.746

**Description of problem:**

Home assistant is unable to refresh firefox after editing any Lovelace card and causes thousands of `Client exceeded max pending messages [2]: 512` errors in a very short time.

**Expected behaviour:**
Home assistant is able to refresh firefox correctly or display a static page requesting a refresh without generating 100s of errors per second. 

**Relevant config:**

**Steps to reproduce this problem:**

1. Open and edit any pre-existing Lovelace card in Firefox.
2. Close the card edit and exit UI edit mode, do not refresh the browser.
3. Navigate to another view or sidebar menu a few times.
4. Sooner rather than later Firefox starts throwing up rapidly repeating “Refresh Lovelace” toast notifications in the lower left corner. This may eventually cause the screen to show a blank page with a useless “Refresh Lovelace” button in the centre but the toast notification loop and error generation continues.

Standard browser refresh clears the error loop.

**Javascript errors shown in the web inspector (if applicable):**

```

```

**Additional information:**

**Comment by al1138** [comment]
Same issue here..

**Comment by bramkragten** [comment]
Can't replicate this, any custom cards in use?

**Comment by al1138** [comment]
Yes, many :)

**Comment by bramkragten** [comment]
Can you try it without them and see if the problem is still there?

**Comment by al1138** [comment]
I'm not 100% sure but I think it's https://github.com/thomasloven/lovelace-auto-entities.
I have added it today, just after the 104 upgrade. I remove it and it's seems to be ok.
Thanks

**Comment by zappasmart** [comment]
Same problem here. Mostly happening when I use lovelace in edit mode.

Though I enabled websocket_api logs at debug level:
```
logger:
  default: info
  logs:
    homeassistant.components.websocket_api: debug
```
analyzing the crazy amount of logs generated in a second is almost impossible.
Wondering if anyone ever found the ultimate solution.

**Comment by astrandb** [comment]
Same problem here. Happens both with Chrome and new Edge. The errors starts when I leave the browser with HA open for a few minutes to hours. The log file home-assistant.log grows rapidly to Gigabyte size. The errors stop immediately when I close the affected browser tab.
Running HA 106.5

**Comment by tomlut** [comment]
Make sure you refresh your browser ( CTRL + Shift =+ R ) or (CTRL +F5) after editing or adding any card.

**Comment by astrandb** [comment]
It is not related to editing anything. It goes crazy after some hours without me doing anything. And, yes, the errors stop after clearing the browser cache.

**Comment by zappasmart** [comment]
> Make sure you refresh your browser ( CTRL + Shift =+ R ) or (CTRL +F5) after editing or adding any card.
> 
> https://community.home-assistant.io/t/client-exceeded-max-pending-messages/158992/5?u=tom_l

I wish it was so easy. I cleared the cache, even used incognito mode. Only after just 30 seconds from logging in the UI was totally unresponsive. I didnt even edit a card or modified anything.

**Comment by zappasmart** [comment]
Browser keeps sending these requests:
![image](https://user-images.githubusercontent.com/53600373/76504753-9ba62000-6448-11ea-8195-0406bd11d00d.png)

Server replies with 101 Switching Protocols:
![image](https://user-images.githubusercontent.com/53600373/76505042-12431d80-6449-11ea-8551-8879d9b854af.png)

**Comment by zsarnett** [comment]
Have you tried this without any custom cards?

**Comment by zappasmart** [comment]
> Have you tried this without any custom cards?

These are the cards installed and used by my UI:
![image](https://user-images.githubusercontent.com/53600373/76962689-f693b780-691f-11ea-9707-4b1c7b54d3fc.png)

I guess I should start progressively uninstalling them one by one and update the lovelace code, until I see any improvement, or do you have any better recommendation on how to troubleshoot?

**Comment by iantrich** [comment]
My money would be on the card templater

**Comment by Mariusthvdb** [comment]
HI Ian, as a side question, but probably related:
I remember reading your remark elsewhere in the community you suggested these max pending messages to be originated from failing jinja templates, is that correct? 
If so, would you know how to debug those messages? The info we now get is as abstract as can be. 
Im stuck in a chicken/egg issue here:
whenever this occurs in the log, I also see disconnects popup, but ont know which causes which... Of course, if the system disconnects, templates can't be evaluated. Otoh, multiple failing templates could cause disconnects? Not sure. Would probably be the former?

Should we have 

{% if not states('entity_to_be_evaluated') == None %}

on all jinja templates then maybe?

**Comment by zappasmart** [comment]
> My money would be on the card templater

1. I was using **iantrich/config-template-card** and the issue was happening.
2. Then I unistalled iantrich/config-template-card and moved all my code to **gadgetchnnel/lovelace-card-templater** but the issue was still there. So I removed gadgetchnnel/lovelace-card-templater, but still with no benefit.
3. Currently, the only templates left are used from **thomasloven/lovelace-card-mod** to style cards based on entity states. My next try will be to remove this and see how it goes.

**Comment by andriej** [comment]
I came here from the clossed issue mentioned above.
Same problem, here are my custom cards:
![image](https://user-images.githubusercontent.com/6303489/77421700-917f0c80-6dcc-11ea-907d-0ed209849578.png)

Maybe it's worth to change issue name, as it's not caused by Firefox (Chrome here).

**Comment by Mariusthvdb** [comment]
yes, nothing to do with the browser, and it has been there for ages....just search the community for max pending messages and prepare ;-)

**Comment by andriej** [comment]
So one more graph that is making me head-scratching:
![image](https://user-images.githubusercontent.com/6303489/77422009-1407cc00-6dcd-11ea-8714-274f5c2dd143.png)

Why does not HA handle those 'clients' caused by the issue, while browser causing the error has been dead-closed? Are those connections open and waiting for... what? Maybe there's some room for improvement in connection handling too (or a bug in sub library maybe).

**Comment by tomlut** [comment]
Actually it was in my case. And I have provided the solution to my issue above. https://github.com/home-assistant/frontend/issues/4404#issuecomment-596306435


If you have a different issue you should open a new one.

**Comment by andriej** [comment]
Great, now I'm stuck between two closed issues. 
@bramkragten could you please reopen the one you find more relevant?

**Comment by andriej** [comment]
@tomlut such a selfish behavior anyway - look at connections counter and think of those browsers you can't refresh (i.e. tablets hang on wall, refrigerators or anything that people have HA on)...
It was NOT firefox issue only.

**Comment by tomlut** [comment]
It was in my case. I use both Firefox and Chrome. Only firefox was exhibiting the behaviour. 

Your fingers obviously aren't broken. Open a new issue with _your symptoms_.

**Comment by andriej** [comment]
I did, as you can see it was closed as a duplicate of your issue: https://github.com/home-assistant/frontend/issues/5328

**Comment by gretel** [comment]
```
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
2021-01-22 16:22:27 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [2875148992] Client exceeded max pending messages [2]: 2048
```

still preset, recent `hass` and `firefox`. if this is closed as being a "browser issue" a potential `denial-of-service` vector is left intact. this should be reopenend.

not saying the discussion above was any good, tho.

**Comment by RolandFran** [comment]
Same issue here. 

Logger: homeassistant.components.websocket_api.http.connection
Source: components/websocket_api/http.py:102
Integration: Home Assistant WebSocket API
First occurred: 23. August 2021, 23:17:28 (57107 occurrences)
Last logged: 10:59:22

[547351624480] Client exceeded max pending messages [2]: 2048
[547351625440] Client exceeded max pending messages [2]: 2048
[547311582080] Client exceeded max pending messages [2]: 2048
[547404090288] Client exceeded max pending messages [2]: 2048
[547355080880] Client exceeded max pending messages [2]: 2048

**Comment by gretel** [comment]
this issue is still present. why is this open for years?

**Comment by ildar170975** [comment]
observed in 2021.11.5
Chrome, Win10x64
HA in Docker

**Comment by michaelblight** [comment]
I've been getting this error for the last couple of days. HA becomes really unstable and semi-restarts - it says it's unavailable but does not fire the "when home assistant is restarted" trigger.

A few suspects: (1) I recently locked down some of the HA userids to local and/or non-admin; (2) I have 4 machines running Glances and integrated to HA, and all 4 were non-responsive - never had this happen before; (3) I recently changed from "lovelace_cast" to "media_player.play_media" for 3 google nests.

I killed HA on my iPad in Chrome (running for weeks/months) as well as my PC, and restarted all the Glances apps. I'm not using Firefox anywhere as others have implicated. Have been error-free for 30 minutes.

It would be very nice if the error listed the IP address. At times the "Client exceeded max pending messages" error is being logged more often than once every 0.01 seconds!!!

<img width="952" alt="image" src="https://user-images.githubusercontent.com/373079/219948861-5e92ce55-3083-4de7-b586-589463e5fa35.png">
